--- a/kernel/nv-linux.h	2015-01-11 05:30:46.000000000 +0100
+++ b/kernel/nv-linux.h	2015-01-25 20:30:32.000000000 +0100
@@ -2000,7 +2000,11 @@
 #if defined(NV_FILE_HAS_INODE)
 #define NV_FILE_INODE(file) (file)->f_inode
 #else
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 19 , 0)
 #define NV_FILE_INODE(file) (file)->f_dentry->d_inode
+#else
+#define NV_FILE_INODE(file) (file)->f_path.dentry->d_inode
+#endif
 #endif
 
 /* Stub out UVM in multi-RM builds */
