--- hprofile-2.0_beta2.orig/extra/rc-scripts/gentoo/hprofile	2014-02-06 18:18:55.000000000 +0000
+++ hprofile-2.0_beta2/extra/rc-scripts/gentoo/hprofile	2014-02-06 18:22:04.000000000 +0000
@@ -4,20 +4,20 @@
 
 
 depend() {
-	before modules
+	after modules
 }
 
 start() {
-	profile=$(/usr/local/sbin/hpdet boot)
-	ebegin "Selecting hardware profile ${profile}"
-	/usr/local/sbin/hprofile boot
+	local vgap=$(/usr/sbin/hpdet vga)
+	ebegin "Selecting hardware profile vga.$vgap"
+	/usr/sbin/hprofile vga
 	eend $?
 }
 
 stop() {
-	profile=$(/usr/local/sbin/hprofile -c boot)
-	ebegin "Stopping hardware profile ${profile}"
-	/usr/local/sbin/hprofile -s boot
+	local vgap=$(/usr/sbin/hprofile -c vga)
+	ebegin "Stopping hardware profile vga.$vgap"
+	/usr/sbin/hprofile -s vga
 	eend $?
 }
 
