--- a/os_dep/linux/wifi_regd.c	2015-04-16 13:32:14.000000000 +0100
+++ b/os_dep/linux/wifi_regd.c	2015-04-19 15:53:26.000000000 +0100
@@ -25,7 +25,7 @@
 
 /* 2G chan 01 - chan 11 */
 #define RTW_2GHZ_CH01_11	\
-	REG_RULE(2412-10, 2462+10, 40, 0, 20, 0)
+	REG_RULE(2412-10, 2462+10, 40, 0, 30, 0)
 
 /*
 *We enable active scan on these a case
@@ -34,12 +34,12 @@
 
 /* 2G chan 12 - chan 13, PASSIV SCAN */
 #define RTW_2GHZ_CH12_13	\
-	REG_RULE(2467-10, 2472+10, 40, 0, 20, \
+	REG_RULE(2467-10, 2472+10, 40, 0, 30, \
 	NL80211_RRF_PASSIVE_SCAN)
 
 /* 2G chan 14, PASSIVS SCAN, NO OFDM (B only) */
 #define RTW_2GHZ_CH14	\
-	REG_RULE(2484-10, 2484+10, 40, 0, 20, \
+	REG_RULE(2484-10, 2484+10, 40, 0, 30, \
 	NL80211_RRF_PASSIVE_SCAN | NL80211_RRF_NO_OFDM)
 
 /* 5G chan 36 - chan 64 */
