--- a/os_dep/linux/usb_intf.c	2015-04-09 02:35:41.000000000 +0100
+++ b/os_dep/linux/usb_intf.c	2015-04-09 02:34:58.000000000 +0100
@@ -299,6 +299,7 @@
 	{USB_DEVICE(0x07B8, 0x8812),.driver_info = RTL8812}, /* Abocom - Abocom */
 	{USB_DEVICE(0x2001, 0x3315),.driver_info = RTL8812}, /* D-Link - Cameo */
 	{USB_DEVICE(0x2001, 0x3316),.driver_info = RTL8812}, /* D-Link - Cameo */
+	{USB_DEVICE(0x20f4, 0x805b),.driver_info = RTL8812}, /* TRENDnet - */
 #endif
 
 #ifdef CONFIG_RTL8821A
