diff -crB a/libgrive/src/drive/State.cc b/libgrive/src/drive/State.cc
*** a/libgrive/src/drive/State.cc	2013-05-02 17:40:04.000000000 +0100
--- b/libgrive/src/drive/State.cc	2014-09-06 08:50:32.000000000 +0100
***************
*** 248,259 ****
  void State::Write( const fs::path& filename ) const
  {
  	Json last_sync ;
! 	last_sync.Add( "sec",	Json(m_last_sync.Sec() ) );
! 	last_sync.Add( "nsec",	Json(m_last_sync.NanoSec() ) );
  	
  	Json result ;
  	result.Add( "last_sync", last_sync ) ;
! 	result.Add( "change_stamp", Json(m_cstamp) ) ;
  	
  	std::ofstream fs( filename.string().c_str() ) ;
  	fs << result ;
--- 248,259 ----
  void State::Write( const fs::path& filename ) const
  {
  	Json last_sync ;
! 	last_sync.Add( "sec",	Json((boost::uint64_t)m_last_sync.Sec() ) );
! 	last_sync.Add( "nsec",	Json((boost::uint64_t)m_last_sync.NanoSec() ) );
  	
  	Json result ;
  	result.Add( "last_sync", last_sync ) ;
! 	result.Add( "change_stamp", Json((boost::uint64_t)m_cstamp) ) ;
  	
  	std::ofstream fs( filename.string().c_str() ) ;
  	fs << result ;
