--- a/drivers/net/wireless/ath/regd.c	2015-04-15 22:49:27.000000000 +0100
+++ b/drivers/net/wireless/ath/regd.c	2015-04-15 23:10:55.000000000 +0100
@@ -32,6 +32,9 @@
  * the flags on our reg_notifier() on a case by case basis.
  */
 
+/* Ignore all regulatory domains if ATH_USER_REGD is enabled */
+#define ATH_USER_REGD 1
+
 /* Only these channels all allow active scan on all world regulatory domains */
 #define ATH9K_2GHZ_CH01_11	REG_RULE(2412-10, 2462+10, 40, 0, 20, 0)
 
