diff -Naur a/src/actions/action.vala b/src/actions/action.vala
--- a/src/actions/action.vala	2014-10-14 18:29:21.000000000 +0200
+++ b/src/actions/action.vala	2016-06-18 18:23:35.721837971 +0200
@@ -78,7 +78,7 @@ public class FileAction : FileActionObject {
 		id = desktop_id;
 		try {
 			kf.load_from_file(desktop_id, 0);
-			FileAction.from_keyfile(kf);
+			this.from_keyfile(kf);
 		}
 		catch(KeyFileError err) {
 		}
@@ -141,7 +141,7 @@ public class FileActionMenu : FileActionObject {
 		id = desktop_id;
 		try {
 			kf.load_from_file(desktop_id, 0);
-			FileActionMenu.from_keyfile(kf);
+			this.from_keyfile(kf);
 		}
 		catch(KeyFileError err) {
 		}
