diff -Naur a/sh/rc-cgroup.sh.in b/sh/rc-cgroup.sh.in
--- a/sh/rc-cgroup.sh.in	2017-03-23 22:50:51.000000000 +0000
+++ b/sh/rc-cgroup.sh.in	2017-05-13 17:18:03.882749932 +0100
@@ -94,7 +94,9 @@
 	openrc_cgroup=/sys/fs/cgroup/openrc
 	if [ -d "$openrc_cgroup" ]; then
 		cgroup="$openrc_cgroup/$RC_SVCNAME"
+		sleep 0.01
 		mkdir -p "$cgroup"
+		sleep 0.01
 		[ -w "$cgroup/tasks" ] && printf "%d" 0 > "$cgroup/tasks"
 	fi
 }
