diff -Naur a/src/wm/meson.build b/src/wm/meson.build
--- a/src/wm/meson.build	2019-03-17 09:06:28.000000000 +0000
+++ b/src/wm/meson.build	2019-05-13 22:24:27.387433478 +0100
@@ -15,8 +15,8 @@
     'wm.vala',
 ]
 
-vapi_mutter = 'libmutter-2'
-dep_mutter = dependency('libmutter-2', version: gnome_minimum_version)
+vapi_mutter = 'libmutter-3'
+dep_mutter = dependency('libmutter-3', version: gnome_minimum_version)
 
 budgie_wm_deps = [
     link_libconfig,
