diff -Naur a/hostapd/config_file.c b/hostapd/config_file.c
--- a/hostapd/config_file.c	2016-10-08 19:25:21.612842189 +0100
+++ b/hostapd/config_file.c	2016-10-08 19:27:45.010843243 +0100
@@ -2863,6 +2863,10 @@
 		}
 #endif /* CONFIG_IEEE80211W */
 #ifdef CONFIG_IEEE80211N
+	} else if (os_strcmp(buf, "noscan") == 0) {
+		conf->noscan = atoi(pos);
+	} else if (os_strcmp(buf, "ht_coex") == 0) {
+		conf->no_ht_coex = !atoi(pos);
 	} else if (os_strcmp(buf, "ieee80211n") == 0) {
 		conf->ieee80211n = atoi(pos);
 	} else if (os_strcmp(buf, "ht_capab") == 0) {
diff -Naur a/src/ap/ap_config.h b/src/ap/ap_config.h
--- a/src/ap/ap_config.h	2016-10-08 19:25:21.621842189 +0100
+++ b/src/ap/ap_config.h	2016-10-08 19:29:23.849843969 +0100
@@ -664,6 +664,8 @@
 
 	int ht_op_mode_fixed;
 	u16 ht_capab;
+	int noscan;
+	int no_ht_coex;
 	int ieee80211n;
 	int secondary_channel;
 	int no_pri_sec_switch;
diff -Naur a/src/ap/hw_features.c b/src/ap/hw_features.c
--- a/src/ap/hw_features.c	2016-10-08 19:25:21.626842189 +0100
+++ b/src/ap/hw_features.c	2016-10-08 19:32:49.579845481 +0100
@@ -474,7 +474,8 @@
 	int ret;
 
 	/* Check that HT40 is used and PRI / SEC switch is allowed */
-	if (!iface->conf->secondary_channel || iface->conf->no_pri_sec_switch)
+	if (!iface->conf->secondary_channel || iface->conf->no_pri_sec_switch ||
+		iface->conf->noscan)
 		return 0;
 
 	hostapd_set_state(iface, HAPD_IFACE_HT_SCAN);
diff -Naur a/src/ap/ieee802_11_ht.c b/src/ap/ieee802_11_ht.c
--- a/src/ap/ieee802_11_ht.c	2016-10-08 19:25:21.628842190 +0100
+++ b/src/ap/ieee802_11_ht.c	2016-10-08 19:35:25.084846623 +0100
@@ -244,6 +244,9 @@
 	if (!(iface->conf->ht_capab & HT_CAP_INFO_SUPP_CHANNEL_WIDTH_SET))
 		return;
 
+	if (iface->conf->noscan || iface->conf->no_ht_coex)
+		return;
+
 	if (len < IEEE80211_HDRLEN + 2 + sizeof(*bc_ie))
 		return;
 
@@ -368,6 +371,9 @@
 	if (iface->current_mode->mode != HOSTAPD_MODE_IEEE80211G)
 		return;
 
+	if (iface->conf->noscan || iface->conf->no_ht_coex)
+		return;
+
 	wpa_printf(MSG_INFO, "HT: Forty MHz Intolerant is set by STA " MACSTR
 		   " in Association Request", MAC2STR(sta->addr));
 
